src/main/scala/com/fulcrumgenomics/basecalling/ExtractBasecallingParams.scala

```diff
@@ -0,0 +1,191 @@
+    val standardLibraryParams = Seq("OUTPUT", "SAMPLE_ALIAS", "LIBRARY_NAME") ++ (if (includeDescription) Seq("DS") else Seq.empty)
+
+    // Headers
+    if (dualIndexed) {
+      barcodeLines += Seq("barcode_sequence_1", "barcode_sequence_2", "barcode_name", "library_name").mkString("\t")
+      libraryParamLines += (Seq("BARCODE_1", "BARCODE_2") ++ standardLibraryParams).mkString("\t")
+    }
+    else {
+      barcodeLines += Seq("barcode_sequence_1", "barcode_name", "library_name").mkString("\t")
+      libraryParamLines += (Seq("BARCODE_1") ++ standardLibraryParams).mkString("\t")
+    }
+
+    // Samples
+    val bams = samples.map { sample =>
+      // Get the library id
+      val libraryId = sample.libraryId.getOrElse(unreachable(s"No library found for sample: ${sample.sampleName}"))
+
+      // Get the barcodes
+      val barcodes = (sample.i7IndexBases, sample.i5IndexBases) match {
+        case (Some(barcode1Bases), Some(barcode2Bases)) => Seq(barcode1Bases, barcode2Bases)
+        case (Some(barcode1Bases), None) => Seq(barcode1Bases)
+        case _ => unreachable(s"No i7 index bases found for sample: ${sample.sampleName}")
+      }
+
+      // Get the description
+      val description = if (includeDescription) Seq(Seq(sample.project, sample.description).flatten.mkString(" - ")) else Seq.empty
+
+      // Add the barcode line
+      barcodeLines += (barcodes ++ Seq(barcodes.mkString, libraryId)).mkString("\t")
+
+      // Add the library param line
+      val bam            = bamFileFrom(output=output, sample=sample, lane=lane)
+      val sampleAlias    = sample.sampleName
+      libraryParamLines += (barcodes ++ Seq(bam, sampleAlias, libraryId) ++ description).mkString("\t")
+      bam
+    }
+    if (bams.toSet.size != bams.size) unreachable("Outputs are not unique when writing the library params file.")
+
+    // Add the unmatched (only for library params!)
+    val description = if (includeDescription) Seq("unmatched") else Seq.empty
+    val barcodes    = if (dualIndexed) Seq("N", "N") else Seq("N")
+    libraryParamLines += (barcodes ++ Seq(output.resolve(s"unmatched.$lane.bam"), "unmatched", "unmatched") ++ description).mkString("\t")
+
+    // Output
+    Io.writeLines(path=barcodeFile, lines=barcodeLines)
+    Io.writeLines(path=libraryParamsFile, lines=libraryParamLines)
+
+    BasecallingParams(barcodeFile=barcodeFile, libraryParamsFile=libraryParamsFile, bams=bams)
+  }
+
+  /** The lane-specific file containing information necessary to run Picard's ExtractIlluminaBarcodes and
+    * CollectIlluminaBasecallingMetrics.
+    * @param output the base output directory.
+    * @param lane the lane number.
+    * @return the path to the barcode file, named "barcode_data.lane-<lane>.txt".
+    */
+  def barcodeFileFrom(output: DirPath, lane: Int): FilePath = {
+    require(0 < lane)
+    output.resolve(s"barcode_params.lane-$lane.txt")
```

Comment:
FWIW when I do this I tend to call these files `barcode-params.$lane.txt`.  Illumina tends to use `L00$lane`.  Not sure what's best, but at a minimum it would make me happier if the names used only one of `-` and `_`!